### PR TITLE
Add ruby 3.3 support and update test suite to use a local server

### PR DESCRIPTION
**Pull request by afromankenobi**
This PR is my attempt to fix Grover's use of Ruby 3.3 and remove the dependency on the external resources the test suite used. It basically:

- Runs a Sinatra app within the test suite that implements most of the features we get from external services.
- Runs the test suite against this internal Sinatra app, allowing testing of Grover's functionality in a controlled environment.
- Includes additional error handling and mock setups in the test suite to address issues encountered during testing.

I hope this is helpful :).

Related to #212 and #213

### CODE REVIEW

**Comment by braindeaf** [comment]
Yep, this is awesome. +1

**Comment by Laykou** [comment]
With Ruby 3.3 I observed weird behavior in the encoding of diacritic (special chars) like:

```
ščťžýáíé
```

I needed to add

`<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">`

**Comment by breisig** [comment]
This is perfect. If it looks good, lets get it merged. :-)

**Comment by afromankenobi** [comment]
@abrom can you take a look at this pls?

File: spec/support/test_app.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    throw(:halt, [401, "Unauthorized access\nYou are denied access to this resource."])
+  end
+
+  def authorized?
+    @auth ||= Rack::Auth::Basic::Request.new(request.env)
+    @auth.provided? && @auth.basic? && @auth.credentials && @auth.credentials == %w[guest guest]
+  end
+end
+
+get '/' do
```
Comment (abrom) [inline spec/support/test_app.rb:20]: nice one 👍 Fixing this has been on my list for months.. this looks like a nice long term solution

File: spec/support/test_app.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  "I'm Feeling Grovery"
+end
+
+get '/cookie_renderer' do
+  cookie_info = request.cookies.map.with_index(1) do |(name, value), i|
+    "#{i}. #{name} #{value}"
```
Comment (abrom) [inline spec/support/test_app.rb:26]: not quite the same rendering as the original, but it'll do (the original used a table)

**Review by abrom — APPROVED** [review]

File: spec/support/test_app.rb (hunk 2 above)
Comment (afromankenobi) [reply spec/support/test_app.rb:26]: Since I wasn't able to get in touch with the original, I had to resort to improvisation.

Comment (abrom) [reply spec/support/test_app.rb:26]: No worries.. I wrote it, I've just "misplaced" the source.. What you did is fine 😄

**Comment by abrom** [comment]
Released in [v1.1.6](https://rubygems.org/gems/grover/versions/1.1.6)

**Pull request merged**